initial ubi8/s2i-base support

danbev opened the pull request:
This commit uses ubi8/s2i-base as the base image. It currently uses rpms
(nodejs/npm) that are in this repository. This is due to some issues we
are having building them. When those issues are resolved the rpms will
be removed and installed from a remote location published by node-rpm.

danbev committed:
initial ubi8/s2i-base support

This commit uses ubi8/s2i-base as the base image. It currently uses rpms
(nodejs/npm) that are in this repository. This is due to some issues we
are having building them. When those issues are resolved the rpms will
be removed and installed from a remote location published by node-rpm.

lholmquist commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+FROM registry.redhat.io/ubi8/s2i-base
```
i noticed that this makes travis fail since it was asking for a login.  i wonder if there is a "public" version somewhere

danbev replied:
Yeah, I'm going to try to find something that does not require a login. I thought it would be available to anyone so there might be a different repository available to pull from.

danbev replied:
Perhaps this is intentional in that we can publish the resulting image (not done for rpms, but we do that for the s2i container) to dockerhub, but the ubi8 base image itself is available in the `registry.access.redhat.com` which requires a login. In this case we need to configure travis with for this. @lance @balunasj Does the above sound right to you or am I missing something?

@lholmquist could you try this locally to verify that the tests pass for you and in the mean time

danbev replied:
Forget that, I'm an idiot. I should have been using `registry.access.redhat.com` instead :(

lance commented on .travis.yml:
```diff
@@ -0,0 +1,16 @@
+before_script:
+  - wget https://github.com/openshift/source-to-image/releases/download/v1.1.10/source-to-image-v1.1.10-27f0729d-linux-amd64.tar.gz
```
There is a 1.1.13 release of `s2i`. Should this maybe use the latest?

danbev replied:
Ah, I did not know that. I just took this as is from the centos-s2i-image repo.

lance commented on build-image-interative.sh:
```diff
@@ -0,0 +1,6 @@
+#export IMAGE="brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhoar-nodejs/nodejs-10:rhoar-nodejs-10-rhel-7-candidate-11567-20180807053735"
+#export IMAGE="brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhoar-nodejs/nodejs-10:rhoar-nodejs-10-rhel-7-containers-candidate-61054-20181001121746"
+export IMAGE="nodeshift/ubi8-s2i-nodejs:12.x"
+docker run -ti -v ${PWD}/test/test-app:/opt/app-root/src $IMAGE /bin/bash
```
What is the purpose of this script? Is it just a convenience?

danbev replied:
That was a mistake, I'll remove it. It's just a script I use for testing.

lance commented on hooks/build:
```diff
@@ -0,0 +1,14 @@
+	--build-arg NODE_VERSION=${NODE_VERSION} \
+	--build-arg NPM_VERSION=${NPM_VERSION} \
+	--build-arg V8_VERSION=${V8_VERSION} \
+	-t ${IMAGE_NAME} .
```
Do you have these hooks enabled in Docker Hub? I had disabled them for the CentOS image builds.

danbev replied:
I've not enabled anything on dockerhub. Can this be deleted?

lance replied:
Yes I believe that everything in `hooks` is for building on Docker Hub. Should probably be removed from `centos7-s2i-nodejs ` as well.

lance replied:
Yes I believe that everything in `hooks` is for building on Docker Hub. Should probably be removed from `centos7-s2i-nodejs ` as well.

danbev committed:
squash: remove build-image-interative.sh

danbev committed:
squash: use 1.1.13 or source-to-image

danbev committed:
remove contents in hooks directory

danbev committed:
squash: minor restructuring

This commit restructures the install_node.sh script so that it will be
easier to share with the productisation build.

danbev committed:
squash: update FROM variables in Makefile

danbev committed:
test: add prod label to test cases

This commit adds a check to also check if the usage label specified in
the Dockerfile matches that of a production build. This makes it easier
for use to just run the same tests without having to modify the tests.

danbev commented:
@lance I've added a few commit which I make it a little easier when doing the internal prod release (to keep these as similar as possible). Could you take another look and see if you think these are ok?

I need to sort out the node-rpm build issue but after that I can update this PR and remove the rpms that I added and instead use the real ones.

lance commented on .gitignore:
```diff
@@ -0,0 +1 @@
+greeting.js
```
Did you add this because [this line](https://github.com/nodeshift/ubi8-s2i-nodejs/pull/1/files#diff-a8b82b354f2e547fabfa7df91cc11b8fR9) generates the file?

danbev replied:
I was added because it was in https://github.com/nodeshift/centos7-s2i-nodejs/blob/master/.gitignore. Note sure it is there.

lance commented on Dockerfile:
```diff
@@ -0,0 +1,54 @@
+      com.redhat.deployments-dir="/opt/app-root/src" \
+      com.redhat.dev-mode="DEV_MODE:false" \
+      com.redhat.dev-mode.port="DEBUG_PORT:5858" \
+      maintainer="Lance Ball <lball@redhat.com>" \
```
It might be good to use multiple labels for `maintainer` here so that others are included (e.g. yourself and Helio).

danbev replied:
Sounds good, I'll do that. I first thought was that having multiple labels with the same key would just overwrite the first one and that only a comma separated list would work.

danbev replied:
I was not able to add multiple labels but a comma separated list looks like this:
```console
$ docker inspect b58ea745d16a | grep maintainer
                "maintainer": "Lance Ball <lball@redhat.com>, Daniel Bevenius <daniel.bevenius@gmail.com>, Helio Frota <00hf11@gmail.com>, Lucas Holmquist <lholmqui@redhat.com>",
```

lance commented on test/test-deploy.sh:
```diff
@@ -0,0 +1,20 @@
+docker push 172.30.1.1:5000/nodeshift/ubi8-s2i-nodejs:12.x
+
+# Creates a new app based on the pushed image.
+oc new-app 172.30.1.1:5000/nodeshift/ubi8-s2i-nodejs:12.x~https://github.com/nodeshift/nodejs-rest-http
```
Is this `172.x.x.x.` IP address always going to be correct?

danbev replied:
I've have no idea if will always be correct. This is just what currently exists  in nodeshift/centos-s2i-nodejs.

lance reviewed (commented):
A couple of minor comments/questions.

danbev committed:
build: add Luke, Helio and myself as maintainers

lance commented:
That works.

On Thu, May 16, 2019 at 9:39 AM Daniel Bevenius <notifications@github.com>
wrote:

> *@danbev* commented on this pull request.
> ------------------------------
>
> In Dockerfile
> <https://github.com/nodeshift/ubi8-s2i-nodejs/pull/1#discussion_r284710895>
> :
>
> > +    DESCRIPTION="Node.js $NODEJS_VERSION available as docker container is a base platform for \
> +building and running various Node.js $NODEJS_VERSION applications and frameworks. \
> +Node.js is a platform built on Chrome's JavaScript runtime for easily building \
> +fast, scalable network applications. Node.js uses an event-driven, non-blocking I/O model \
> +that makes it lightweight and efficient, perfect for data-intensive real-time applications \
> +that run across distributed devices." \
> +    LD_LIBRARY_PATH=/opt/rh/httpd24/root/usr/lib64
> +
> +LABEL io.k8s.description="$DESCRIPTION" \
> +      io.k8s.display-name="Node.js $NODE_VERSION" \
> +      io.openshift.expose-services="8080:http" \
> +      io.openshift.tags="builder,nodejs,nodejs-$NODE_VERSION" \
> +      com.redhat.deployments-dir="/opt/app-root/src" \
> +      com.redhat.dev-mode="DEV_MODE:false" \
> +      com.redhat.dev-mode.port="DEBUG_PORT:5858" \
> +      maintainer="Lance Ball <lball@redhat.com>" \
>
> I was not able to add multiple labels but a comma separated list looks
> like this:
>
> $ docker inspect b58ea745d16a | grep maintainer                "maintainer": "Lance Ball <lball@redhat.com>, Daniel Bevenius <daniel.bevenius@gmail.com>, Helio Frota <00hf11@gmail.com>, Lucas Holmquist <lholmqui@redhat.com>",
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodeshift/ubi8-s2i-nodejs/pull/1?email_source=notifications&email_token=AAAD4UB2USK5BRRRHOV66TDPVVP2HA5CNFSM4HLXZ5DKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBY22UGA#discussion_r284710895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAD4UFQWHPUR4HWR4VEVO3PVVP2HANCNFSM4HLXZ5DA>
> .
>


-- 

Lance Ball

He/Him

Principal Software Engineer

Red Hat <https://www.redhat.com/>

Twitter: @lanceball

danbev committed:
squash: use proper rpms from node-rpm

lance reviewed (approved):
LGTM

danbev commented:
Landed in abb7563, 06bfc54, f960e31, d22e7bd, d28aea8.
